lldb/source/Core/Debugger.cpp

```diff
@@ -1433,11 +1434,30 @@ void Debugger::SetDestroyCallback(
 static void PrivateReportProgress(Debugger &debugger, uint64_t progress_id,
                                   std::string title, std::string details,
                                   uint64_t completed, uint64_t total,
-                                  bool is_debugger_specific) {
+                                  bool is_debugger_specific,
+                                  uint32_t progress_broadcast_bit) {
   // Only deliver progress events if we have any progress listeners.
   const uint32_t event_type = Debugger::eBroadcastBitProgress;
-  if (!debugger.GetBroadcaster().EventTypeHasListeners(event_type))
+  const uint32_t category_event_type = Debugger::eBroadcastBitProgressCategory;
+  if (!debugger.GetBroadcaster().EventTypeHasListeners(event_type |
+                                                       category_event_type))
     return;
+
+  if (debugger.GetBroadcaster().EventTypeHasListeners(category_event_type)) {
+    ProgressManager &progress_manager = ProgressManager::Instance();
+    auto map_refcount = progress_manager.GetProgressCategoryMap().lookup(title);
+
+    // Only broadcast the event to the progress category bit if it's an initial
+    // or final report for that category. Since we're broadcasting for the
+    // category specifically, clear the details.
+    if (progress_broadcast_bit == Debugger::eBroadcastBitProgressCategory) {
+      EventSP event_sp(new Event(
+          category_event_type,
+          new ProgressEventData(progress_id, std::move(title), "", completed,
+                                total, is_debugger_specific)));
+      debugger.GetBroadcaster().BroadcastEvent(event_sp);
+    }
+  }
```

Comment:
I think it would be better from a layering perspective to keep `PrivateReportProgress` "dumb" and pass it the right arguments from the ProgressManager. The map held by the progress manager should be a private implementation detail. By letting the debugger mess with it you lose the ability to synchronize access to it. When we do the timeouts, this isn't going to work.

I expect this function to simply broadcast to whichever broadcast bit is set.

Reply:
So IIUC, instead of passing the broadcast bit as an argument we just set it in the progress manager and access it from there in the Debugger?

Reply:
Not really. I'm saying that I think we don't need to store it at all, because its implied in when we call `DebuggerReportProgress`.

Can ProgressManager call `ReportProgress(bit = eBroadcastBitProgress | eBroadcastBitProgressCategory)` for the first and last update, and every other time call `ReportProgress(bit = eBroadcastBitProgress) every other time?

Reply:
So that instead of having it be a member of the struct or the class we just call `Debugger::ReportProgress(<other information>, eBroadcastBitProgress | eBroadcastBitProgressCategory)` and set the bit as needed from the progress manager itself?

Reply:
Such that we set `progress_category_bit` from the progress manager and broadcast to that as such:
```      EventSP event_sp(new Event(
          category_event_type,
          new ProgressEventData(progress_id, std::move(title), "", completed,
                                total, is_debugger_specific)));
      debugger.GetBroadcaster().BroadcastEvent(event_sp);
```

Reply:
Also looking back at `PrivateReportProgress`, the variables for the instance and refcount are leftover from a previous implementation that I had, they shouldn't be here anymore.